src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -361,134 +369,266 @@ public class SqlMutationEngine : IMutationEngine
                 // Records affected tells us that item was successfully deleted.
                 // No records affected happens for a DELETE request on nonexistent object
                 if (resultProperties is not null
-                    && resultProperties.TryGetValue(nameof(DbDataReader.RecordsAffected), out object? value)
-                    && Convert.ToInt32(value) > 0)
+                    && resultProperties.TryGetValue(nameof(DbDataReader.RecordsAffected), out object? value))
                 {
+                    // DbDataReader.RecordsAffected contains the number of rows changed deleted. 0 if no records were deleted.
+                    // When the flow reaches this code block and the number of records affected is 0, then it means that no failure occurred at the database layer
+                    // and that the item identified by the specified PK was not found.
+                    if (Convert.ToInt32(value) == 0)
+                    {
+                        string prettyPrintPk = "<" + string.Join(", ", context.PrimaryKeyValuePairs.Select(kv_pair => $"{kv_pair.Key}: {kv_pair.Value}")) + ">";
+
+                        throw new DataApiBuilderException(
+                            message: $"Could not find item with {prettyPrintPk}",
+                            statusCode: HttpStatusCode.NotFound,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.ItemNotFound);
+                    }
+
                     return new NoContentResult();
                 }
             }
-            else if (context.OperationType is EntityActionOperation.Upsert || context.OperationType is EntityActionOperation.UpsertIncremental)
+            else
             {
-                DbResultSet? upsertOperationResult = null;
+                string roleName = GetHttpContext().Request.Headers[AuthorizationResolver.CLIENT_ROLE_HEADER];
+                bool isReadPermissionConfiguredForRole = _authorizationResolver.AreRoleAndOperationDefinedForEntity(context.EntityName, roleName, EntityActionOperation.Read);
+                bool isDatabasePolicyDefinedForReadAction = false;
+                JsonDocument? selectOperationResponse = null;
```

Comment:
Add the `using` here.

Reply:
Initially, I had tried `using`.  

Only read-only variables (and they have to be initialized when declaration) can be used with `using` statements. Here, we want to be able to assign the results from the database results at a later point. 

So, added a `finally` block to dispose.